### PR TITLE
[mlir][vector] Add a pattern to fuse extract(constant_mask)

### PR DESCRIPTION
**Pull request by Hsiangkai**
This pattern will rewrite
ExtractOp(ConstantMaskOp) -> ConstantMaskOp
or
ExtractOp(ConstantMaskOp) -> Constant

### CODE REVIEW

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 1]
```diff
@@ -2039,6 +2039,60 @@ class ExtractOpFromCreateMask final : public OpRewritePattern<ExtractOp> {
   }
 };
 
+// Pattern to rewrite
+// ExtractOp(ConstantMaskOp) -> ConstantMaskOp
+// or
+// ExtractOp(ConstantMaskOp) -> ConstantOp
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2045]: Could you elaborate on when one rewrite is used vs the other?

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 2]
```diff
@@ -2039,6 +2039,60 @@ class ExtractOpFromCreateMask final : public OpRewritePattern<ExtractOp> {
...
+        extractOp.getVector().getDefiningOp<vector::ConstantMaskOp>();
+    if (!constantMaskOp)
+      return failure();
+
+    auto maskDimSizes = constantMaskOp.getMaskDimSizes().getValue();
+    auto resultTy = extractOp.getResult().getType();
+    if (resultTy.isa<mlir::VectorType>()) {
+      auto resultVectorTy = resultTy.cast<mlir::VectorType>();
+      auto resultRank = resultVectorTy.getRank();
+      auto n = maskDimSizes.size();
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2062]: nit: spell out `auto` when the type is not explicit in the code.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 3]
```diff
@@ -2039,6 +2039,60 @@ class ExtractOpFromCreateMask final : public OpRewritePattern<ExtractOp> {
...
+      std::vector<int64_t> indices;
+      for (size_t i = n - resultRank; i < n; i++)
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2064]: nit: use pre-increment per coding standards

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 4]
```diff
@@ -2039,6 +2039,60 @@ class ExtractOpFromCreateMask final : public OpRewritePattern<ExtractOp> {
...
+      return success();
+    } else if (resultTy.isa<mlir::IntegerType>()) {
+      // Extract a scalar. All indices must be static.
+      ArrayRef<int64_t> extractOpPos = extractOp.getStaticPosition();
+      if (maskDimSizes.size() != extractOpPos.size())
+        return failure();
+
+      auto boolType = rewriter.getI1Type();
+      auto setAttr = IntegerAttr::get(boolType, 1);
+      for (size_t i = 0; i < extractOpPos.size(); i++) {
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2080]: nit: move loop upper-bound to variable (e.g., `size_t i = 0, end = extractOpPos.size(); i < end...`) per coding guidelines

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 5]
```diff
@@ -2039,6 +2039,60 @@ class ExtractOpFromCreateMask final : public OpRewritePattern<ExtractOp> {
...
+        if (cast<IntegerAttr>(maskDimSizes[i]).getInt() <= extractOpPos[i]) {
+          setAttr = IntegerAttr::get(boolType, 0);
+          break;
+        }
+      }
+
+      rewriter.replaceOpWithNewOp<arith::ConstantOp>(extractOp, boolType,
+                                                     setAttr);
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2088]: Some comments about what this code is doing would be helpful

File: mlir/test/Dialect/Vector/canonicalize.mlir [hunk 6]
```diff
@@ -2567,3 +2567,50 @@ func.func @load_store_forwarding_rank_mismatch(%v0: vector<4x1x1xf32>, %arg0: te
...
+// CHECK-LABEL: func.func @extract_from_create_mask() -> i1 {
+func.func @extract_from_create_mask() -> i1 {
+// CHECK:      %[[TRUE:.*]] = arith.constant true
+// CHECK-NEXT: return %[[TRUE]] : i1
+  %c2 = arith.constant 2 : index
+  %c3 = arith.constant 3 : index
+  %mask = vector.create_mask %c2, %c2, %c3 : vector<4x4x4xi1>
+  %extract = vector.extract %mask[1, 1, 2] : i1 from vector<4x4x4xi1>
+  return %extract : i1
+}
```
Comment (dcaballe) [inline mlir/test/Dialect/Vector/canonicalize.mlir:2604]: Could you add a few tests where the indices are not static?

**Review by dcaballe — COMMENTED** [review]
Thanks for the contribution! LG! A few nit comments, mostly.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 1 above)
Comment (Hsiangkai) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp]: Done.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 2 above)
Comment (Hsiangkai) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp]: Done.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 3 above)
Comment (Hsiangkai) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp]: Done.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 4 above)
Comment (Hsiangkai) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp]: Done.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 5 above)
Comment (Hsiangkai) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp]: Done.

File: mlir/test/Dialect/Vector/canonicalize.mlir (hunk 6 above)
Comment (Hsiangkai) [reply mlir/test/Dialect/Vector/canonicalize.mlir:2604]: Done.

**Comment by Hsiangkai** [comment]
Ping.

File: mlir/test/Dialect/Vector/canonicalize.mlir [hunk 7]
```diff
@@ -2567,3 +2567,74 @@ func.func @load_store_forwarding_rank_mismatch(%v0: vector<4x1x1xf32>, %arg0: te
...
+
+// -----
+
+// CHECK-LABEL: func.func @extract_subvector_with_dynamic_positions
+func.func @extract_subvector_with_dynamic_positions(%index: index) -> vector<6xi1> {
+// CHECK:      %[[S0:.*]] = vector.constant_mask [4] : vector<6xi1>
+// CHECK-NEXT: return %[[S0]] : vector<6xi1>
+  %mask = vector.constant_mask [2, 3, 4] : vector<4x5x6xi1>
+  %extract = vector.extract %mask[1, %index] : vector<6xi1> from vector<4x5x6xi1>
```
[the remark this answers is not in the document]
Comment (Hsiangkai) [reply mlir/test/Dialect/Vector/canonicalize.mlir:2640]: You are right. Fixed.

**Review by dcaballe — APPROVED** [review]
Thanks!

**Comment by kuhar** [comment]
@Hsiangkai sorry for not reviewing this earlier, I just came back from vacation. Could you please address the issues above?

**Comment by dcaballe** [comment]
@qedawkins, could you add the test case that failed here, please?

**Comment by hanhanW** [comment]
I think it is embedded in the revert commit message. Let me copy it to here:

This pattern is producing incorrect IR. For example,

```mlir
func.func @extract_subvector_from_constant_mask() -> vector<16xi1> {
  %mask = vector.constant_mask [2, 3] : vector<16x16xi1>
  %extract = vector.extract %mask[8] : vector<16xi1> from vector<16x16xi1>
  return %extract : vector<16xi1>
}
```

Canonicalizes to

```mlir
func.func @extract_subvector_from_constant_mask() -> vector<16xi1> {
  %0 = vector.constant_mask [3] : vector<16xi1>
  return %0 : vector<16xi1>
}
```

Where it should be a zero mask because the extraction index (8) is
greater than the constant mask size along that dim (2).

**Comment by qedawkins** [comment]
ah, sorry. Yeah what Hanhan posted is the example test case.

**Pull request merged**